reporter/collector_reporter.go

```diff
@@ -0,0 +1,491 @@
+	log "github.com/sirupsen/logrus"
+	"go.opentelemetry.io/collector/consumer/consumerprofiles"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/pprofile"
+	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
+
+	"go.opentelemetry.io/ebpf-profiler/libpf"
+	"go.opentelemetry.io/ebpf-profiler/libpf/xsync"
+)
+
+// Assert that we implement the full Reporter interface.
+var _ Reporter = (*CollectorReporter)(nil)
+
+// OTLPReporter receives and transforms information to be Collector Collector compliant.
+type CollectorReporter struct {
+	cfg          *Config
+	nextConsumer consumerprofiles.Profiles
+
+	// stopSignal is the stop signal for shutting down all background tasks.
+	stopSignal chan struct{}
+
+	// hostmetadata stores metadata that is sent out with every request.
+	hostmetadata *lru.SyncedLRU[string, string]
+
+	// executables stores metadata for executables.
+	executables *lru.SyncedLRU[libpf.FileID, execInfo]
+
+	// frames maps frame information to its source location.
+	frames *lru.SyncedLRU[libpf.FileID, *xsync.RWMutex[map[libpf.AddressOrLineno]sourceInfo]]
+
+	// traceEvents stores reported trace events (trace metadata with frames and counts)
+	traceEvents xsync.RWMutex[map[traceAndMetaKey]*traceEvents]
+
+	// samplesPerSecond is the number of samples per second.
+	samplesPerSecond int
+}
+
+// NewCollector builds a new CollectorReporter
+func NewCollector(cfg *Config, nextConsumer consumerprofiles.Profiles) (*CollectorReporter, error) {
+	executables, err :=
+		lru.NewSynced[libpf.FileID, execInfo](cfg.ExecutablesCacheElements, libpf.FileID.Hash32)
+	if err != nil {
+		return nil, err
+	}
+	executables.SetLifetime(1 * time.Hour) // Allow GC to clean stale items.
+
+	frames, err := lru.NewSynced[libpf.FileID,
+		*xsync.RWMutex[map[libpf.AddressOrLineno]sourceInfo]](
+		cfg.FramesCacheElements, libpf.FileID.Hash32)
+	if err != nil {
+		return nil, err
+	}
+	frames.SetLifetime(1 * time.Hour) // Allow GC to clean stale items.
+
+	// Next step: Dynamically configure the size of this LRU.
+	// Currently, we use the length of the JSON array in
+	// hostmetadata/hostmetadata.json.
+	hostmetadata, err := lru.NewSynced[string, string](115, hashString)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CollectorReporter{
+		cfg:          cfg,
+		nextConsumer: nextConsumer,
+
+		executables:  executables,
+		frames:       frames,
+		hostmetadata: hostmetadata,
+		traceEvents:  xsync.NewRWMutex(map[traceAndMetaKey]*traceEvents{}),
+
+		samplesPerSecond: cfg.SamplesPerSecond,
+	}, nil
+}
+
+func (r *CollectorReporter) Start(context.Context) error {
+	go func() {
+		tick := time.NewTicker(r.cfg.ReportInterval)
+		defer tick.Stop()
+		for {
+			select {
+			case <-r.stopSignal:
+				return
+			case <-tick.C:
+				if err := r.reportProfile(context.Background()); err != nil {
+					log.Errorf("Request failed: %v", err)
+				}
+				tick.Reset(libpf.AddJitter(r.cfg.ReportInterval, 0.2))
+			}
```

Comment:
There was a recent change in `otlp_reporter.go`, please consider applying it here as well: https://github.com/open-telemetry/opentelemetry-ebpf-profiler/blob/38669a76c9b630260d627f224967ab1f2fef41b8/reporter/otlp_reporter.go#L392